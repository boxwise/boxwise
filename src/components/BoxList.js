import React from "react";
import { Link } from "react-router-dom";
import { withStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import Divider from "@material-ui/core/Divider";
import Typography from "@material-ui/core/Typography";

import Progress from "./Progress.js";
import ProductSelect from "./ProductSelect";

const styles = theme => ({
  root: {
    marginTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 4
  },
  empty: theme.mixins.gutters({
    textAlign: "center",
    marginTop: theme.spacing.unit * 4
  }),
  progress: {
    margin: theme.spacing.unit * 4,
    display: "flex",
    justifyContent: "center"
  },
  productSelect: theme.mixins.gutters({})
});

const BoxList = ({
  classes,
  isLoading,
  boxes,
  products,
  selectedProductFilter,
  onChangeProductFilter
}) => {
  const getProduct = id => {
    const result = products.filter(product => product.id === id);
    if (result[0]) {
      return result[0];
    }
    return {
      category: "[none]",
      name: "[none]"
    };
  };

  let list;

  if (isLoading) {
    list = (
      <div className={classes.progress}>
        <Progress />
      </div>
    );
  } else if (!boxes.length) {
    list = (
      <Typography className={classes.empty}>
        No boxes. <Link to="/">Make some?</Link>
      </Typography>
    );
  } else {
    list = (
      <List>
        <Divider />
        {boxes.map(box => (
          <React.Fragment key={box.id}>
            <ListItem>
              <ListItemText
                primary={
                  <span>
                    <strong>{box.humanID}</strong> {box.quantity}x{" "}
                    {getProduct(box.product.id).category} /{" "}
                    {getProduct(box.product.id).name}
                  </span>
                }
                secondary={box.comment ? box.comment : "Â "}
              />
            </ListItem>
            <Divider />
          </React.Fragment>
        ))}
      </List>
    );
  }

  return (
    <div className={classes.root}>
      <div className={classes.productSelect}>
        <ProductSelect
          products={products}
          value={selectedProductFilter}
          onChange={onChangeProductFilter}
        />
      </div>
      <br />
      {list}
    </div>
  );
};

export default withStyles(styles)(BoxList);
