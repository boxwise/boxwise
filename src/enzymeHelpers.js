export { mount, shallow } from "enzyme";

export const setInputFieldValue = (component, name, value) => {
  component.find(`input[name='${name}']`).simulate("change", {
    target: { name: `${name}`, value: `${value}` },
    persist: () => {}
  });
};
