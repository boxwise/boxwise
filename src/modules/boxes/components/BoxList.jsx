import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { withStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import Divider from "@material-ui/core/Divider";
import Typography from "@material-ui/core/Typography";

import { ProductSelect } from "modules/products/components";
import Progress from "components/Progress";

const styles = theme => ({
  root: {
    marginTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 4
  },
  empty: theme.mixins.gutters({
    textAlign: "center",
    marginTop: theme.spacing.unit * 4
  }),
  progress: {
    margin: theme.spacing.unit * 4,
    display: "flex",
    justifyContent: "center"
  },
  productSelect: theme.mixins.gutters({})
});

export const BoxList = ({
  classes,
  boxesWithProductInfo,
  products,
  fetchBoxesAndProducts
}) => {
  const [filterByProductId, setFilterByProductId] = useState(undefined);

  useEffect(() => {
    fetchBoxesAndProducts();
  }, [fetchBoxesAndProducts]);
  let filteredData = boxesWithProductInfo.data;
  if (filterByProductId)
    filteredData = filteredData.filter(
      box => box.productId === filterByProductId
    );

  let list;

  if (boxesWithProductInfo.loading) {
    list = (
      <div className={classes.progress}>
        <Progress />
      </div>
    );
  } else if (!filteredData.length) {
    list = (
      <Typography className={classes.empty}>
        No boxes{filterByProductId ? " that matches your current filter" : ""}.{" "}
        <Link to="/">Make some?</Link>
      </Typography>
    );
  } else {
    list = (
      <List>
        <Divider />
        {filteredData.map(box => (
          <React.Fragment key={box.id}>
            <ListItem>
              <ListItemText
                primary={
                  <span>
                    <strong>{box.humanId}</strong> {box.quantity}x{" "}
                    {box.productCategory} / {box.productName}
                  </span>
                }
                secondary={box.comment ? box.comment : "Â "}
              />
            </ListItem>
            <Divider />
          </React.Fragment>
        ))}
      </List>
    );
  }

  return (
    <div className={classes.root}>
      <div className={classes.productSelect}>
        <ProductSelect
          products={products.data}
          value={filterByProductId}
          onChange={value => setFilterByProductId(value)}
        />
      </div>
      <br />
      {list}
    </div>
  );
};

export default withStyles(styles)(BoxList);
