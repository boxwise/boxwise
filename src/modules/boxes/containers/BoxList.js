import { connect } from "react-redux";

import { getAllProductsFromState } from "modules/products/reducer";
import { getAllProducts } from "modules/products/actions";

import { fetchBoxes } from "../actions";
import BoxList from "../components/BoxList";
import { getBoxesWithProductInfoFromState } from "../reducer";

export default connect(
  state => ({
    boxesWithProductInfo: getBoxesWithProductInfoFromState(state),
    products: getAllProductsFromState(state)
  }),
  dispatch => ({
    fetchBoxesAndProducts: () =>
      dispatch(fetchBoxes()).then(dispatch(getAllProducts()))
  })
)(BoxList);
