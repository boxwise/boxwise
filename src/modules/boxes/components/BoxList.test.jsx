import React from "react";

import { shallow } from "enzymeHelpers";

import { BoxList } from "./BoxList";

describe("BoxList", () => {
  const emptyProducts = { data: [] };
  const emptyBoxesWithProductInfo = { loading: false, data: [] };
  const emptyFetch = () => {};

  it("renders correctly with no boxes", () => {
    const wrapper = shallow(
      <BoxList
        classes={{ titleText: "styles" }}
        products={emptyProducts}
        boxesWithProductInfo={emptyBoxesWithProductInfo}
        fetchBoxesAndProducts={emptyFetch}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly with when loading", () => {
    const wrapper = shallow(
      <BoxList
        classes={{ titleText: "styles" }}
        products={emptyProducts}
        boxesWithProductInfo={{ loading: true, data: [] }}
        fetchBoxesAndProducts={emptyFetch}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when boxes present", () => {
    const boxes = [
      { id: "xx", humanId: "123", productName: "TShirt", quantity: 100 },
      { id: "xy", humanId: "456", productName: "Shorts", quantity: 29 }
    ];
    const wrapper = shallow(
      <BoxList
        classes={{ titleText: "styles" }}
        products={{ data: [] }}
        boxesWithProductInfo={{ data: boxes }}
        fetchBoxesAndProducts={emptyFetch}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when no boxes present matching filter", () => {
    const boxes = [
      {
        id: "xy",
        humanId: "xyz",
        productName: "TShirt",
        productId: "abc",
        quantity: 100
      }
    ];
    const wrapper = shallow(
      <BoxList
        classes={{ titleText: "styles" }}
        products={{ data: [] }}
        boxesWithProductInfo={{ data: boxes }}
        fetchBoxesAndProducts={emptyFetch}
      />
    );
    wrapper.find("ProductSelect").simulate("change", "xyz");
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when boxes match filter", () => {
    const boxes = [
      {
        id: "xy",
        humanId: "xyz",
        productName: "TShirt",
        productId: "abc",
        quantity: 100
      }
    ];
    const wrapper = shallow(
      <BoxList
        classes={{ titleText: "styles" }}
        products={{ data: [] }}
        boxesWithProductInfo={{ data: boxes }}
        fetchBoxesAndProducts={emptyFetch}
      />
    );
    wrapper.find("ProductSelect").simulate("change", "abc");
    expect(wrapper).toMatchSnapshot();
  });
});
