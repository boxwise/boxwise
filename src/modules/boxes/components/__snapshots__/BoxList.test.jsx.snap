// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxList renders correctly when boxes match filter 1`] = `
<div>
  <div>
    <ProductSelect
      onChange={[Function]}
      products={Array []}
      value="abc"
    />
  </div>
  <br />
  <WithStyles(List)>
    <WithStyles(Divider) />
    <WithStyles(ListItem)>
      <WithStyles(ListItemText)
        primary={
          <span>
            <strong>
              xyz
            </strong>
             
            100
            x
             
             / 
            TShirt
          </span>
        }
        secondary=" "
      />
    </WithStyles(ListItem)>
    <WithStyles(Divider) />
  </WithStyles(List)>
</div>
`;

exports[`BoxList renders correctly when boxes present 1`] = `
<div>
  <div>
    <ProductSelect
      onChange={[Function]}
      products={Array []}
    />
  </div>
  <br />
  <WithStyles(List)>
    <WithStyles(Divider) />
    <WithStyles(ListItem)>
      <WithStyles(ListItemText)
        primary={
          <span>
            <strong>
              123
            </strong>
             
            100
            x
             
             / 
            TShirt
          </span>
        }
        secondary=" "
      />
    </WithStyles(ListItem)>
    <WithStyles(Divider) />
    <WithStyles(ListItem)>
      <WithStyles(ListItemText)
        primary={
          <span>
            <strong>
              456
            </strong>
             
            29
            x
             
             / 
            Shorts
          </span>
        }
        secondary=" "
      />
    </WithStyles(ListItem)>
    <WithStyles(Divider) />
  </WithStyles(List)>
</div>
`;

exports[`BoxList renders correctly when no boxes present matching filter 1`] = `
<div>
  <div>
    <ProductSelect
      onChange={[Function]}
      products={Array []}
      value="xyz"
    />
  </div>
  <br />
  <WithStyles(Typography)>
    No boxes
     that matches your current filter
    .
     
    <Link
      to="/"
    >
      Make some?
    </Link>
  </WithStyles(Typography)>
</div>
`;

exports[`BoxList renders correctly with no boxes 1`] = `
<div>
  <div>
    <ProductSelect
      onChange={[Function]}
      products={Array []}
    />
  </div>
  <br />
  <WithStyles(Typography)>
    No boxes
    .
     
    <Link
      to="/"
    >
      Make some?
    </Link>
  </WithStyles(Typography)>
</div>
`;

exports[`BoxList renders correctly with when loading 1`] = `
<div>
  <div>
    <ProductSelect
      onChange={[Function]}
      products={Array []}
    />
  </div>
  <br />
  <div>
    <Progress />
  </div>
</div>
`;
